include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+                               "Domain>::Tvmult_add called"));
+    };
+
+    reinit_range_vector = [](Range &, bool)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::reinit_range_vector method called"));
+    };
+
+    reinit_domain_vector = [](Domain &, bool)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::reinit_domain_vector method called"));
+    };
+  }
+
+  /**
+   * Create a LinearOperator object from an object @p op for which the
+   * conversion function <code>linop</code> is defined.
+   */
+  template<typename Op>
+  LinearOperator (const Op &op)
+  {
+    *this = linop<Range, Domain, Op>(op);
+  }
+
+  /**
+   * Copy assignment operator for an object @p op for which the conversion
+   * function <code>linop</code> is defined.
+   */
+  template <typename Op>
+  LinearOperator<Range, Domain> &operator=(const Op &op)
+  {
+    return *this = linop<Range, Domain, Op>(op);
+  }
+
+  /**
+   * Application of the LinearOperator object to a vector u of the @tref
+   * Domain space giving a vector v of the @tref Range space.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult;
+
+  /**
+   * Application of the LinearOperator object to a vector u of the @tref
+   * Domain space. The result is added to the vector v.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult_add;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector u of
+   * the @tref Range space giving a vector v of the @tref Domain
+   * space.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector @p u of
+   * the @tref Range space.The result is added to the vector @p v.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult_add;
+
+  /**
+   * Initializes a vector v of the Range space to be directly usable
+   * as the destination parameter in an application of vmult. Similar to
+   * the reinit functions of the vector classes, the boolean determines
+   * whether a fast initalization is done, i.e., if it is set to false the
+   * content of the vector is set to 0.
+   */
+  std::function<void(Range &v, bool fast)> reinit_range_vector;
+
+  /**
+   * Initializes a vector of the Domain space to be directly usable as the
+   * source parameter in an application of vmult. Similar to the reinit
+   * functions of the vector classes, the boolean determines whether a fast
+   * initalization is done, i.e., if it is set to false the content of the
+   * vector is set to 0.
+   */
+  std::function<void(Domain &v, bool fast)> reinit_domain_vector;
+
+  /**
+   * A memory object for vectors of @tref Range space used for intermediate
+   * storage during computations in @ref vmult, etc.
+   */
+  mutable GrowingVectorMemory<Range> range_vector_memory;
```

Comment:
`GrowingVectorMemory` is for all practical purposes a singleton as it draws vectors from a global pool. I think it would be reasonable if you let every function that needs such an object have a static local variable (or even non-static, if you want), instead of making it a variable at class scope.